staging/src/k8s.io/apiextensions-apiserver/test/integration/versioning_test.go

```diff
@@ -136,3 +140,266 @@ func testStoragedVersionInCRDStatus(t *testing.T, ns string, noxuDefinition *api
+// 2. Add a version = v1beta2 with served = true. Ensure that both v1beta1 and v1beta2 appear in discovery
+// and both APIs are usable by checking that the old CR still has version = v1beta1 and creating a new CR with version = v1beta2.
+// 3. Set v1beta2 with stored = true. Create a new CR and ensure that it can persist with version = v1beta2
+// and update the CR with version = v1beta1 to v1beta2, so that all CRs now have version = v1beta2.
+// 4. Set v1beta1 with served = false, ensure that it disappears from discovery
+// and the API is no longer callable by attempting to create a CR with version = v1beta1.
+// 5. Remove v1beta1 from crd.Status.StoredVersions.
+// 6. Remove v1beta1 from crd.Spec.Versions and set crd.Spec.Version as v1beta2.
+func TestConvertVersion(t *testing.T) {
+	stopCh, apiExtensionClient, dynamicClient, err := testserver.StartDefaultServerWithClients()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer close(stopCh)
+
+	// create CRD with version = v1beta1
+	noxuDefinition := testserver.NewNoxuCustomResourceDefinition(apiextensionsv1beta1.NamespaceScoped)
+	noxuDefinition, err = testserver.CreateNewCustomResourceDefinition(noxuDefinition, apiExtensionClient, dynamicClient)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// ensure v1beta1 appears in discovery
+	group := "mygroup.example.com"
+	if err := wait.Poll(100*time.Millisecond, wait.ForeverTestTimeout, func() (done bool, err error) {
+		_, err = apiExtensionClient.Discovery().ServerResourcesForGroupVersion(group + "/v1beta1")
+		if err == nil {
+			return true, nil
+		}
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}); err != nil {
+		t.Fatalf("failed to see v1beta1 in discovery: %v", err)
+	}
+
+	// create a CR and make sure it's version is v1beta1
+	ns := "not-the-default"
+	noxuV1beta1ResourceClient := newNamespacedCustomResourceClient(ns, dynamicClient, noxuDefinition)
+	noxuInstanceToCreate := testserver.NewNoxuInstance(ns, "foo")
+	createdFooInstance, err := noxuV1beta1ResourceClient.Create(noxuInstanceToCreate)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if createdFooInstance.GetAPIVersion() != "mygroup.example.com/v1beta1" {
+		t.Fatalf("invalid version: expected \"mygroup.example.com/v1beta1\", got %v", createdFooInstance.GetAPIVersion())
+	}
+
+	// add v1beta2 with served = true and storage = false
+	gottenCRD, err := apiExtensionClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(noxuDefinition.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("could not get CRD: %v", err)
+	}
+	v1beta2 := apiextensionsv1beta1.CustomResourceDefinitionVersion{
+		Name:    "v1beta2",
+		Served:  true,
+		Storage: false,
+	}
+	gottenCRD.Spec.Versions = append(gottenCRD.Spec.Versions, v1beta2)
+	err = wait.Poll(500*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
+		updatedCRD, err := apiExtensionClient.ApiextensionsV1beta1().CustomResourceDefinitions().Update(gottenCRD)
```

Comment:
You don't need a pool-retry  to update CRD in the test. The assumption is nobody else is working on this CRD as this is a single thread test and also update will be immediately available. If you want you can retry update until it succeeded but when it succeed, the Spec.Versions will be 2 (and if it is not, it is a bug, should assert on it instead of retry).

Reply:
:+1:

> If you want you can retry update until it succeeded but when it succeed

I remember a problem with not polling for this in some other tests. I'll try without polling once again and confirm.